src/Microsoft.AspNetCore.Sockets.Client/ServerSentEventsTransport.cs

```diff
@@ -0,0 +1,232 @@
+            try
+            {
+                while (await _application.Input.WaitToReadAsync(cancellationToken))
+                {
+                    messages = new List<SendMessage>();
+                    while (!cancellationToken.IsCancellationRequested && _application.Input.TryRead(out SendMessage message))
+                    {
+                        messages.Add(message);
+                    }
+
+                    if (messages.Count > 0)
+                    {
+                        _logger.LogDebug("Sending {0} message(s) to the server using url: {1}", messages.Count, sendUrl);
+
+                        var request = new HttpRequestMessage(HttpMethod.Post, sendUrl);
+                        request.Headers.UserAgent.Add(DefaultUserAgentHeader);
+
+                        var memoryStream = new MemoryStream();
+
+                        var pipe = memoryStream.AsPipelineWriter();
+                        var output = new PipelineTextOutput(pipe, TextEncoder.Utf8);
+                        await WriteMessagesAsync(messages, output, MessageFormat.Binary);
+
+                        memoryStream.Seek(0, SeekOrigin.Begin);
+
+                        request.Content = new StreamContent(memoryStream);
+                        request.Content.Headers.ContentType = MediaTypeHeaderValue.Parse(MessageFormatter.GetContentType(MessageFormat.Binary));
+
+                        var response = await _httpClient.SendAsync(request);
+                        response.EnsureSuccessStatusCode();
+
+                        _logger.LogDebug("Message(s) sent successfully");
+                        foreach (var message in messages)
+                        {
+                            message.SendResult?.TrySetResult(null);
+                        }
+                    }
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                _logger.LogError("Send cancelled");
+
+                if (messages != null)
+                {
+                    foreach (var message in messages)
+                    {
+                        message.SendResult?.TrySetCanceled();
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogDebug("Error while sending to '{0}' : '{1}'", sendUrl, ex);
+                if (messages != null)
+                {
+                    foreach (var message in messages)
+                    {
+                        message.SendResult?.TrySetException(ex);
+                    }
+                }
+                throw;
+            }
+            finally
+            {
+                // Make sure the poll loop is terminated
+                _transportCts.Cancel();
+            }
+
+            _logger.LogInformation("Send loop stopped");
+        }
+
+        private async Task WriteMessagesAsync(IList<SendMessage> messages, PipelineTextOutput output, MessageFormat format)
```

Comment:
No, use `List<T>`. For private code, it's best to move the concrete `List<T>` around so that when you a) avoid interface dispatch and b) when you enumerate you can use the `struct` enumerator that `List<T>` provides, rather than boxing it.